design/src/docs/pokemon_service.md

```diff
@@ -0,0 +1,239 @@
+A `Router` is a tower `Service` that routes requests to the implemented services; hence it [implements](https://github.com/awslabs/smithy-rs/blob/db48039065bec890ef387385773b37154b555b14/rust-runtime/aws-smithy-http-server/src/routing/mod.rs#L302) `Service`
+like the other operations.
+
+To re-use code and not to overload a router with possibly unneeded operations, it also [implements](https://github.com/awslabs/smithy-rs/blob/db48039065bec890ef387385773b37154b555b14/rust-runtime/aws-smithy-http-server/src/routing/mod.rs#L146)
+the `tower::Layer` trait. Middleware are added as layers. The Pokémon example uses them:
+```rust
+let shared_state = Arc::new(State::default());
+let app = app.layer(
+    ServiceBuilder::new()
+        .layer(TraceLayer::new_for_http())
+        .layer(AddExtensionLayer::new(shared_state)),
+);
+```
+The service is run by `hyper`:
+```rust
+hyper::Server::bind(&bind).serve(app.into_make_service());
+```
+
+
+### Folder structure
+Before describing how shapes are generated, it is helpful to know the folder structure of smithy-rs.
+Code generation happens in:
+* `/codegen`: it contains shared code for both client and server, but only generates a client
+* `/codegen-server`: server only. This project started with `codegen` to generate a client, but client and server share common code; that code lives in `codegen`, which `codegen-server` depends on
+* `/aws`: the AWS SDK, it deals with AWS services specifically. The folder structure reflects the project's, with the `rust-runtime` and the `codegen`
+* `/rust-runtime`: the generated client and server crates may depend on crates in this folder. Crates here are not code generated
+
+`/rust-runtime` crates ("runtime crates") are added to a crate's dependency only when used. If a model uses event streams, it will depend on `aws-smithy-eventstream`.
+
+### Generating code
+`smithy-rs` is a gradle plugin, not a command. Its entry point is in [RustCodegenPlugin::execute](https://github.com/awslabs/smithy-rs/blob/db48039065bec890ef387385773b37154b555b14/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/RustCodegenPlugin.kt#L34) and
+inherits from `SmithyBuildPlugin` in [smithy-build](https://github.com/awslabs/smithy/tree/main/smithy-build). Code generation is in Kotlin and shared common, non-Rust specific code with the `smithy` Java repository.
+
+The comment at the beginning of `execute` described what a `Decorator` is and uses the following terms:
+* Context: contains the model being generated, projection and settings for the build
+* Decorator: (also referred to as customizations) customizes how code is being generated. AWS services are required to sign with the SigV4 protocol, and [a decorator](https://github.com/awslabs/smithy-rs/blob/db48039065bec890ef387385773b37154b555b14/aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4SigningDecorator.kt#L45) adds Rust code to sign requests and responses.
```

Comment:
Technically, a decorator can comprise _multiple_ customizations. See for example `RequiredCustomizations.kt`. It's a decorator that just applies a collection of customizations.

Customizations are really code injections/personalizations you can perform in certain sections of the code generation. See for example inheritors of `NamedSectionGenerator`.